boa_engine/src/builtins/function/mod.rs

```diff
@@ -473,40 +473,56 @@ impl BuiltInFunctionObject {
         new_target: &JsValue,
         args: &[JsValue],
         r#async: bool,
+        generator: bool,
         context: &mut Context,
     ) -> JsResult<JsObject> {
-        let prototype =
-            get_prototype_from_constructor(new_target, StandardConstructors::function, context)?;
+        let default = if r#async {
+            StandardConstructors::async_function
+        } else if generator {
+            StandardConstructors::generator_function
+        } else {
+            StandardConstructors::function
+        };
+
+        let prototype = get_prototype_from_constructor(new_target, default, context)?;
         if let Some((body_arg, args)) = args.split_last() {
-            let parameters =
-                if args.is_empty() {
-                    FormalParameterList::empty()
-                } else {
-                    let mut parameters = Vec::with_capacity(args.len());
-                    for arg in args {
-                        parameters.push(arg.to_string(context)?);
+            let parameters = if args.is_empty() {
+                FormalParameterList::empty()
+            } else {
+                let mut parameters = Vec::with_capacity(args.len());
+                for arg in args {
+                    parameters.push(arg.to_string(context)?);
+                }
+                let mut parameters = parameters.join(",");
```

Comment:
It seems a bit too much work here, right? Why do we need to re-convert the parameters to a string?

Especially because I see that we later call the parser. Would it be possible to have a shorthand function in the parser that doesn't need a raw `(...)` and can just work with the vector?

Reply:
We for sure need to convert every argument to a string individually, because this is of course valid: 
```javascript
const arg = Object("a")
typeof arg // 'object'
const f = Function(arg, "console.log(a)")
f(1) // 1
```

We probably could have a special parser for this, but tbh I don't think it's that imporant. I think (hope) that this is not used that much in production code. I will open an issue to possibly implement this if it does not cause too much code duplication in the parser.

Reply:
See https://github.com/boa-dev/boa/issues/2180